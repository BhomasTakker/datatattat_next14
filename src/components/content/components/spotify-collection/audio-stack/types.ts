import { SearchType } from "@/types/api/spotify";
import { CollectionItem } from "@/types/data-structures/collection/item/item";

export type SpotifyCollectionItem = CollectionItem & {
	id: string;
	media: {
		type: SearchType;
	};
};

export type AudioStackClientProps = {
	items: SpotifyCollectionItem[];
};

export type SpotifyArticleProps = {
	item: SpotifyCollectionItem;
	onClick: (index: number) => void;
	index: number;
};
