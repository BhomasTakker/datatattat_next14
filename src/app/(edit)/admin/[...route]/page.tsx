import { getServerSession } from "next-auth";
import styles from "../../../page.module.scss";
import { EditPage } from "@/components/edit/edit-page";
import { getUserById } from "@/lib/mongo/actions/user";
import { Session } from "@/types/auth/session";
import { options } from "@/app/api/auth/[...nextauth]/options";
import { PATHS } from "@/lib/routing/paths";

export const dynamic = "force-dynamic";

type PageProps = {
	params: Promise<{ route: string[] }>;
};

export default async function Page({ params }: PageProps) {
	const session = (await getServerSession(options)) as Session;
	const { user: sessionUser } = session;
	const { role } = await getUserById(sessionUser.user_id);

	// need select from multiple levels of admin?
	const adminLevel = PATHS.home();

	const { route } = await params;
	const joined = route.join("/");

	// Get/Check level(s) of access
	if (role !== "admin") {
		return <div>Unauthorized</div>;
	}

	return (
		<div className={styles.page}>
			<EditPage route={`/${joined}`} title="Admin Edit" isAdminEdit={true} />
		</div>
	);
}
