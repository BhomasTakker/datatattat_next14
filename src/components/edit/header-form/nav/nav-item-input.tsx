import { useFormContext } from "react-hook-form";
import styles from "./nav-item-input.module.scss";
import { FaArrowUp, FaArrowDown } from "react-icons/fa";
import { MdDelete } from "react-icons/md";

type Link = {
	label: string;
	route: string;
};

export const NavItemInput = ({
	link,
	index,
	onMove,
	onDelete,
}: {
	link: Link;
	index: number;
	onMove: (index: number, direction: string) => void;
	onDelete: (index: number) => void;
}) => {
	const { label, route } = link;
	const { register } = useFormContext();

	const labelInputId = `header-${index}-label`;
	const routeInputId = `header-${index}-route`;

	return (
		<div className={styles.root}>
			<div className={styles.inputs}>
				<div className={styles.input}>
					<label htmlFor={labelInputId}>Label</label>
					<input
						id={labelInputId}
						{...register(labelInputId, {
							required: true,
							maxLength: 20,
							minLength: 2,
						})}
						type="text"
						defaultValue={label}
						required
					/>
				</div>
				<div className={styles.input}>
					<label htmlFor={routeInputId}>Route</label>
					<input
						id={routeInputId}
						{...register(routeInputId, { required: true, minLength: 2 })}
						type="text"
						defaultValue={route}
						required
					/>
				</div>
			</div>
			<div className={styles.controls}>
				<div className={styles.icon} onClick={() => onMove(index, "up")}>
					<FaArrowUp />
				</div>
				<div className={styles.icon} onClick={() => onMove(index, "down")}>
					<FaArrowDown />
				</div>
				<div className={styles.icon} onClick={() => onDelete(index)}>
					<MdDelete />
				</div>
			</div>
		</div>
	);
};
