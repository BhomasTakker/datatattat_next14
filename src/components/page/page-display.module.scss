@use "../../scss/abstracts/units/" as *;

.page {
	display: flex;
	flex-direction: column;
}
