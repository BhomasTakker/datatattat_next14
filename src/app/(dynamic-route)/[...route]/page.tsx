import { getPage } from "@/actions/page/page-actions";
import { IPage } from "@/types/page";
import { PageDisplay } from "@/components/page/page-display";
import styles from "../../page.module.scss";
import { generateMetaDataFromPage } from "@/lib/metadata/generate-metadata";
import { initialiseServices } from "@/lib/services/intialise-services";

type Params = Promise<{ route: string[] }>;
type Props = {
	params: Params;
};

export const revalidate = 960;

// export async function generateStaticParams() {
// 	return [
// 		{ route: ["world"] },
// 		{ route: ["us"] },
// 		{ route: ["uk"] },
// 		{ route: ["ukraine"] },
// 		{ route: ["trade-war"] },
// 	];
// }

export const generateMetadata = async ({ params }: Props) => {
	await initialiseServices();
	const { route } = await params;
	const joined = route.join("/");

	const metadata = await generateMetaDataFromPage(`/${joined}`);
	return metadata;
};

export default async function Page({ params }: Props) {
	await initialiseServices();
	const { route } = await params;
	const joined = route.join("/");

	const page = (await getPage(`/${joined}`)) as IPage;

	if (!page) {
		// omg do something
		return <h1>404</h1>;
	}

	return (
		<div className={styles.page}>
			<PageDisplay page={page} />
		</div>
	);
}
