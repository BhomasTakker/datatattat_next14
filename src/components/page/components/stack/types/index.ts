import { PageStackCollectionVariants } from "@/types/components/page/stack";

export type VerticalStackProps = {
	variant: PageStackCollectionVariants.Vertical;
};

export const ContainerWidth = {
	XXS: "180",
	XS: "240",
	SM: "320",
	MD: "640",
	LG: "768",
	XL: "1024",
	XXL: "1280",
} as const;

export const ContainerWidthOptions = {
	XXS: "XXS",
	XS: "XS",
	SM: "SM",
	MD: "MD",
	LG: "LG",
	XL: "XL",
	XXL: "XXL",
} as const;

type ContainerWidth = (typeof ContainerWidth)[keyof typeof ContainerWidth];

export type ContainerWidthOptions = keyof typeof ContainerWidth;

export const ContainerHeight = {
	XXS: "240",
	XS: "360",
	SM: "480",
	MD: "600",
	LG: "720",
	XL: "840",
	XXL: "960",
} as const;

export const ContainerHeightOptions = {
	XXS: "XXS",
	XS: "XS",
	SM: "SM",
	MD: "MD",
	LG: "LG",
	XL: "XL",
	XXL: "XXL",
} as const;

type ContainerHeight = (typeof ContainerHeight)[keyof typeof ContainerHeight];

type ContainerHeightOptions = keyof typeof ContainerHeight;
export const containerWidthOptions: ContainerWidthOptions[] = Object.keys(
	ContainerWidth
) as ContainerWidthOptions[];

// Row configuration for row-stack layout
export type Row = {
	columns: number;
	evenColumns: boolean;
	// you wouldn't max/min height - you would set a height
	maxHeight: ContainerHeightOptions;
	minWidth: ContainerWidthOptions;
	// these are all actually strings!
	// Unless we can/want to convert these after http requests use strings
	index?: string; // optional for default row - specifies which row this config applies to
};

export type Column = {
	minWidth: ContainerWidthOptions;
	maxHeight: ContainerHeightOptions;
	evenColumns: boolean;
};

export type RowStackProps = {
	variant: PageStackCollectionVariants.RowStack;
	defaultRow: Row; // Default configuration used for any row not specifically configured
	rows?: Row[]; // Optional array of row-specific configurations - index property determines which row each config applies to
};

export type PageStackProps = {
	variant: PageStackCollectionVariants;
} & (VerticalStackProps | RowStackProps);
