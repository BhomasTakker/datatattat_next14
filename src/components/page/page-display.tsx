import { IPage } from "@/types/page";
import { PageProfile } from "./profile/page-profile";
import { PageComponentFactory } from "./components/page-component-factory";
import styles from "./page-display.module.scss";
import { getCurrentRoute } from "@/utils/route";

export const PageDisplay = async ({ page }: { page: IPage }) => {
	const { content, meta, profile } = page || {};

	// not here
	const currentRoute = await getCurrentRoute();
	const isUserRoute = currentRoute?.includes("/users/") || false;
	const isClient = isUserRoute;

	return (
		<main className={styles.page}>
			{/* REPLACE with meta function */}
			{/* Page head for meta data, seo, etc */}
			{/* <PageHead headData={meta} /> */}
			{/* Page profile data for title, description, createdBy, etc */}
			<PageProfile profile={profile} />
			<PageComponentFactory content={content} isClient={isClient} />
		</main>
	);
};
