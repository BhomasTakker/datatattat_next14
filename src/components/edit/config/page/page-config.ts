import { GenericInput } from "@/types/edit/inputs/inputs";
import { EditInputs } from "../../inputs/inputs";
import { COMPONENT_CONFIG } from "../components/component-config";
import { PAGE_GRID_CONFIG } from "./components/page-grid";
import { PAGE_STACK_CONFIG } from "./components/stack/page-stack";

enum pageContainers {
	STACK = "Stack",
	GRID = "Grid",
}

type pageContainersProps = typeof PAGE_STACK_CONFIG | typeof PAGE_GRID_CONFIG;

const pageContainersMap = new Map<string, pageContainersProps>([
	[pageContainers.STACK, PAGE_STACK_CONFIG],
	[pageContainers.GRID, PAGE_GRID_CONFIG],
]);

export const PAGE_CONFIG: GenericInput = {
	// id used will be content
	id: "page",
	type: EditInputs.inputList,
	label: "Page",

	inputs: [
		{
			// add button here to save as page template - no we would want to save meta and page properties too no?
			// add button here to demo page
			id: "title",
			type: EditInputs.title,
			title: "Add Page interaction buttons here",
		},
		{
			id: "containerType",
			type: EditInputs.objectSelect,
			label: "Page Container",
			defaultValue: pageContainers.STACK,
			required: true,
			options: [pageContainers.STACK],
			optionMap: pageContainersMap,
			// save to/create on props object
			optionId: "props",
		},
		COMPONENT_CONFIG,
	],
};
