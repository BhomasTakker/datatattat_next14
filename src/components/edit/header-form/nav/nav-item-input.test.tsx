import React from "react";
import { render, screen, fireEvent } from "@testing-library/react";
import { useForm, FormProvider } from "react-hook-form";
import { NavItemInput } from "./nav-item-input";
import { IPage } from "@/types/page";

// Mock styles import
jest.mock("./nav-item-input.module.scss", () => ({
	root: "root",
	inputs: "inputs",
	input: "input",
	controls: "controls",
	icon: "icon",
}));

const pages: IPage[] = [
	// @ts-expect-error - mock data
	{ route: "/home", id: "1", title: "Home" },
	// @ts-expect-error - mock data
	{ route: "/about", id: "2", title: "About" },
];

const link = { label: "Test Label", route: "/about" };

const renderWithForm = (props = {}) => {
	const Wrapper: React.FC<React.PropsWithChildren<{}>> = ({ children }) => {
		const methods = useForm();
		return <FormProvider {...methods}>{children}</FormProvider>;
	};
	return render(
		<NavItemInput
			link={link}
			index={0}
			onMove={jest.fn()}
			onDelete={jest.fn()}
			pages={pages}
			{...props}
		/>,
		{ wrapper: Wrapper }
	);
};

describe("NavItemInput", () => {
	it("renders label and route inputs", () => {
		renderWithForm();
		expect(screen.getByLabelText(/Label/i)).toBeInTheDocument();
		expect(screen.getByLabelText(/Route/i)).toBeInTheDocument();
	});

	it("renders the correct default label value", () => {
		renderWithForm();
		const input = screen.getByLabelText(/Label/i) as HTMLInputElement;
		expect(input.value).toBe("Test Label");
	});

	it("renders all page options in the select", () => {
		renderWithForm();
		const select = screen.getByLabelText(/Route/i) as HTMLSelectElement;
		expect(select).toBeInTheDocument();
		expect(screen.getByText("Select a route")).toBeInTheDocument();
		expect(screen.getByText("/home")).toBeInTheDocument();
		expect(screen.getByText("/about")).toBeInTheDocument();
	});

	it("selects the correct route option", () => {
		renderWithForm();
		const select = screen.getByLabelText(/Route/i) as HTMLSelectElement;
		expect(select.value).toBe("/about");
	});

	it("calls onMove with 'up' when up arrow is clicked", () => {
		const onMove = jest.fn();
		renderWithForm({ onMove });
		const element = screen.getByTestId("move-up");
		fireEvent.click(element); // Up arrow
		expect(onMove).toHaveBeenCalledWith(0, "up");
	});

	it("calls onMove with 'down' when down arrow is clicked", () => {
		const onMove = jest.fn();
		renderWithForm({ onMove });
		const element = screen.getByTestId("move-down");
		fireEvent.click(element); // Down arrow
		expect(onMove).toHaveBeenCalledWith(0, "down");
	});

	it("calls onDelete when delete icon is clicked", () => {
		const onDelete = jest.fn();
		renderWithForm({ onDelete });
		const element = screen.getByTestId("delete");
		fireEvent.click(element); // Delete icon
		expect(onDelete).toHaveBeenCalledWith(0);
	});

	describe("Snaps", () => {
		it("renders NavItemInput unchanged", async () => {
			const { container } = renderWithForm();
			expect(container).toMatchSnapshot();
		});
	});
});
