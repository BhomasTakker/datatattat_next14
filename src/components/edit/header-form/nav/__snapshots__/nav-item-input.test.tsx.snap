// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`NavItemInput Snaps renders NavItemInput unchanged 1`] = `
<div>
  <div
    class="root"
  >
    <div
      class="inputs"
    >
      <div
        class="input"
      >
        <label
          for="header-0-label"
        >
          Label
        </label>
        <input
          id="header-0-label"
          name="header-0-label"
          required=""
          type="text"
          value="Test Label"
        />
      </div>
      <div
        class="input"
      >
        <label
          for="header-0-route"
        >
          Route
        </label>
        <select
          id="header-0-route"
          name="header-0-route"
        >
          <option
            value=""
          >
            Select a route
          </option>
          <option
            value="/home"
          >
            /home
          </option>
          <option
            value="/about"
          >
            /about
          </option>
        </select>
      </div>
    </div>
    <div
      class="controls"
    >
      <div
        aria-label="Move Up"
        aria-role="button"
        class="icon"
        data-testid="move-up"
      >
        <svg
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 448 512"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M34.9 289.5l-22.2-22.2c-9.4-9.4-9.4-24.6 0-33.9L207 39c9.4-9.4 24.6-9.4 33.9 0l194.3 194.3c9.4 9.4 9.4 24.6 0 33.9L413 289.4c-9.5 9.5-25 9.3-34.3-.4L264 168.6V456c0 13.3-10.7 24-24 24h-32c-13.3 0-24-10.7-24-24V168.6L69.2 289.1c-9.3 9.8-24.8 10-34.3.4z"
          />
        </svg>
      </div>
      <div
        aria-label="Move Down"
        aria-role="button"
        class="icon"
        data-testid="move-down"
      >
        <svg
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 448 512"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M413.1 222.5l22.2 22.2c9.4 9.4 9.4 24.6 0 33.9L241 473c-9.4 9.4-24.6 9.4-33.9 0L12.7 278.6c-9.4-9.4-9.4-24.6 0-33.9l22.2-22.2c9.5-9.5 25-9.3 34.3.4L184 343.4V56c0-13.3 10.7-24 24-24h32c13.3 0 24 10.7 24 24v287.4l114.8-120.5c9.3-9.8 24.8-10 34.3-.4z"
          />
        </svg>
      </div>
      <div
        aria-label="Delete"
        aria-role="button"
        class="icon"
        data-testid="delete"
      >
        <svg
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M0 0h24v24H0z"
            fill="none"
          />
          <path
            d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6v12zM19 4h-3.5l-1-1h-5l-1 1H5v2h14V4z"
          />
        </svg>
      </div>
    </div>
  </div>
</div>
`;
